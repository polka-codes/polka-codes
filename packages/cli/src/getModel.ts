import { createAnthropic } from '@ai-sdk/anthropic'
import { createDeepSeek } from '@ai-sdk/deepseek'
import { createOpenAI } from '@ai-sdk/openai'
import type { LanguageModelV2 } from '@ai-sdk/provider'
import { createOpenRouter } from '@openrouter/ai-sdk-provider'
import { createOllama } from 'ollama-ai-provider-v2'

export enum AiProvider {
  Anthropic = 'anthropic',
  Ollama = 'ollama',
  DeepSeek = 'deepseek',
  OpenRouter = 'openrouter',
  OpenAI = 'openai',
}

export type ModelConfig = {
  provider: AiProvider
  model: string
  apiKey?: string
  baseUrl?: string
}

export const getModel = (config: ModelConfig, debugLogging = false): LanguageModelV2 => {
  const fetchOverride: typeof fetch = debugLogging
    ? ((async (url, options) => {
        // log outgoing request
        console.log('-> Request URL:', url)
        console.log('-> Request Headers:', options?.headers)
        console.log('-> Request Body:', JSON.parse(options?.body as any))

        const res = await fetch(url, options)
        console.log('<- Response Status:', res.status)

        const contentType = res.headers.get('content-type') || ''
        // if it's a streaming response, clone the body stream
        if (contentType.includes('text/event-stream') && res.body) {
          const [branch, clientStream] = res.body.tee()
          // consume branch to log chunks
          ;(async () => {
            const reader = branch.getReader()
            const decoder = new TextDecoder()
            let done = false
            while (!done) {
              const { value, done: d } = await reader.read()
              done = d
              if (value) {
                const text = decoder.decode(value)
                console.log('<- Stream chunk:', text.replace(/\n/g, '\\n'))
              }
            }
          })()
          // return the other branch to the SDK
          return new Response(clientStream, {
            headers: res.headers,
            status: res.status,
          })
        }

        // non-stream: read and log full body
        const full = await res.text()
        console.log('<- Response Body:')
        console.dir(JSON.parse(full), { depth: null })
        return new Response(full, {
          headers: res.headers,
          status: res.status,
        })
      }) as typeof fetch)
    : fetch

  switch (config.provider) {
    case AiProvider.Anthropic: {
      const anthropic = createAnthropic({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })
      return anthropic(config.model)
    }

    case AiProvider.Ollama: {
      const ollama = createOllama({
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })
      return ollama(config.model)
    }

    case AiProvider.DeepSeek: {
      const deepseek = createDeepSeek({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })
      return deepseek(config.model)
    }

    case AiProvider.OpenRouter: {
      const openrouter = createOpenRouter({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })
      return openrouter.chat(config.model, {
        usage: { include: true },
      })
    }

    case AiProvider.OpenAI: {
      const openai = createOpenAI({
        apiKey: config.apiKey,
        baseURL: config.baseUrl,
        fetch: fetchOverride,
      })

      return openai(config.model)
    }
  }
}
