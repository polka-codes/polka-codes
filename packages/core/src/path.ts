/**
 * Returns the directory portion of a path string.
 * Strips trailing slashes, then takes everything up to the last slash.
 */
export function dirname(path: string): string {
  if (path.length === 0) return '.'
  const isRooted = path[0] === '/'
  // Trim trailing slashes (but leave root "/")
  let end = path.length - 1
  while (end > 0 && path[end] === '/') end--
  const idx = path.lastIndexOf('/', end)
  if (idx < 0) {
    return isRooted ? '/' : '.'
  }
  if (isRooted && idx === 0) {
    return '/'
  }
  return path.slice(0, idx)
}

/**
 * Normalizes a path by resolving "." and ".." segments.
 */
function normalize(path: string): string {
  const isAbsolute = path.startsWith('/')
  const segments = path.split('/').filter(Boolean)
  const stack: string[] = []
  for (const seg of segments) {
    if (seg === '.') continue
    if (seg === '..') {
      if (stack.length && stack[stack.length - 1] !== '..') {
        stack.pop()
      } else if (!isAbsolute) {
        stack.push('..')
      }
    } else {
      stack.push(seg)
    }
  }
  let result = stack.join('/')
  if (!result && !isAbsolute) return '.'
  if (result && path.endsWith('/')) result += '/'
  return (isAbsolute ? '/' : '') + result
}

/**
 * Joins all given path segments, then normalizes the result.
 */
export function join(...parts: string[]): string {
  if (parts.length === 0) return '.'
  let combined = ''
  for (const p of parts) {
    if (typeof p !== 'string') {
      throw new TypeError('Arguments to join must be strings')
    }
    if (p) {
      combined = combined ? `${combined}/${p}` : p
    }
  }
  return normalize(combined)
}
