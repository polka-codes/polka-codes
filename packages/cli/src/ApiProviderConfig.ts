import type { Config } from '@polka-codes/cli-shared'
import type { ToolFormat } from '@polka-codes/core'
import { AiProvider } from './getModel'

const isSonnet4 = (model: string) => model.includes('claude-sonnet-4')

const getToolFormat = (model: string, toolFormat?: ToolFormat) => {
  if (toolFormat) {
    return toolFormat
  }
  if (isSonnet4(model)) {
    return 'native'
  }
  return 'polka-codes'
}

const defaultModels = {
  [AiProvider.Anthropic]: 'claude-sonnet-4-20250514',
  [AiProvider.Ollama]: 'deepseek-r1:32b',
  [AiProvider.DeepSeek]: 'deepseek-chat',
  [AiProvider.OpenRouter]: 'google/gemini-2.5-pro',
  [AiProvider.OpenAI]: 'gpt-4.1',
  [AiProvider.GoogleVertex]: 'gemini-2.5-pro',
}

export class ApiProviderConfig {
  readonly defaultProvider?: AiProvider
  readonly providers: Readonly<Partial<Record<AiProvider, { apiKey: string; defaultModel?: string; defaultParameters: any }>>>
  readonly commands?: Partial<Record<string, { provider?: AiProvider; model?: string; parameters: any; toolFormat?: ToolFormat }>>
  readonly agents?: Partial<Record<string, { provider?: AiProvider; model?: string; parameters: any; toolFormat?: ToolFormat }>>
  readonly defaultParameters: any
  readonly toolFormat?: ToolFormat

  constructor(config: Config) {
    this.defaultProvider = config.defaultProvider as AiProvider | undefined
    this.defaultParameters = config.defaultParameters ?? {}
    this.providers = config.providers ?? {}
    this.commands = config.commands as Partial<
      Record<string, { provider?: AiProvider; model?: string; parameters: any; toolFormat?: ToolFormat }>
    >
    this.agents = config.agents as Partial<
      Record<string, { provider?: AiProvider; model?: string; parameters: any; toolFormat?: ToolFormat }>
    >
    this.toolFormat = config.toolFormat
  }

  getConfigForCommand(command: string) {
    // TODO: strong type command
    const { provider, model, parameters, toolFormat } = this.commands?.[command] ?? this.commands?.default ?? {}
    const finalProvider = provider ?? this.defaultProvider
    if (!finalProvider) {
      return undefined
    }
    const { apiKey, defaultModel, defaultParameters, location, project, keyFile } = (this.providers[finalProvider] as any) ?? {}
    const finalModel = model ?? defaultModel ?? defaultModels[finalProvider]
    const finalParameters = {
      ...this.defaultParameters,
      ...(defaultParameters ?? {}),
      ...(parameters ?? {}),
    }
    const finalToolFormat = getToolFormat(finalModel, toolFormat ?? this.toolFormat)
    return {
      provider: finalProvider,
      model: finalModel,
      apiKey,
      location,
      project,
      keyFile,
      parameters: finalParameters,
      toolFormat: finalToolFormat,
    }
  }

  getConfigForAgent(agent: string) {
    // TODO: strong type agent
    const { provider, model, parameters, toolFormat } = this.agents?.[agent] ?? this.agents?.default ?? {}
    const finalProvider = provider ?? this.defaultProvider
    if (!finalProvider) {
      return undefined
    }
    const { apiKey, defaultModel, defaultParameters, location, project, keyFile } = (this.providers[finalProvider] as any) ?? {}
    const finalModel = model ?? defaultModel ?? defaultModels[finalProvider]
    const finalParameters = {
      ...this.defaultParameters,
      ...(defaultParameters ?? {}),
      ...(parameters ?? {}),
    }
    const finalToolFormat = getToolFormat(finalModel, toolFormat ?? this.toolFormat)
    return {
      provider: finalProvider,
      model: finalModel,
      apiKey,
      location,
      project,
      keyFile,
      parameters: finalParameters,
      toolFormat: finalToolFormat,
    }
  }
}
