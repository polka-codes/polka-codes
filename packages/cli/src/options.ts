// Generated by polka.codes

import os from 'node:os'
import { AiServiceProvider } from '@polka-codes/core'
import type { Command } from 'commander'
import { set } from 'lodash'

import { ApiProviderConfig } from './ApiProviderConfig'
import { loadConfig } from './config'
import { type Env, getEnv } from './env'

export interface CliOptions {
  config?: string | string[]
  apiProvider?: string
  model?: string
  apiKey?: string
  maxMessageCount?: number
  budget?: number
  verbose?: number
  baseDir?: string
  agent?: string
}

export function addSharedOptions(command: Command) {
  return command
    .option('-c --config <paths>', 'Path to config file(s)', (value: string, prev: string[]) => prev.concat(value), [])
    .option('--api-provider <provider>', 'API provider')
    .option('--model <model>', 'Model ID')
    .option('--api-key <key>', 'API key')
    .option('--max-messages <iterations>', 'Maximum number of messages to send. Default to 50', Number.parseInt, 50)
    .option('--budget <budget>', 'Budget for the AI service. Default to $10', Number.parseFloat)
    .option('-v --verbose', 'Enable verbose output. Use -v for level 1, -vv for level 2', (value, prev) => prev + 1, 0)
    .option('-d --base-dir <path>', 'Base directory to run commands in')
    .option('--agent <agent>', 'Initial agent to use (default: architect)')
}

export function parseOptions(options: CliOptions, cwdArg?: string, home: string = os.homedir(), env: Partial<Env> = getEnv()) {
  let cwd = cwdArg
  if (options.baseDir) {
    process.chdir(options.baseDir)
    cwd = options.baseDir
    console.log('Changed working directory to', cwd)
  } else {
    cwd = process.cwd()
  }

  const config = loadConfig(options.config, cwd, home) ?? {}

  const defaultProvider = (options.apiProvider || env.POLKA_API_PROVIDER || config.defaultProvider) as AiServiceProvider | undefined
  const defaultModel = options.model || env.POLKA_MODEL || config.defaultModel

  if (defaultProvider && defaultModel) {
    set(config, ['providers', defaultProvider, 'defaultModel'], defaultModel)
  }

  const apiKey = options.apiKey || env.POLKA_API_KEY

  if (apiKey) {
    if (!defaultProvider) {
      throw new Error('Must specify a provider if providing an API key')
    }
    set(config, ['providers', defaultProvider, 'apiKey'], apiKey)
  }

  // Set provider-specific API keys
  if (env.ANTHROPIC_API_KEY) {
    set(config, ['providers', AiServiceProvider.Anthropic, 'apiKey'], env.ANTHROPIC_API_KEY)
  }
  if (env.DEEPSEEK_API_KEY) {
    set(config, ['providers', AiServiceProvider.DeepSeek, 'apiKey'], env.DEEPSEEK_API_KEY)
  }
  if (env.OPENROUTER_API_KEY) {
    set(config, ['providers', AiServiceProvider.OpenRouter, 'apiKey'], env.OPENROUTER_API_KEY)
  }

  const providerConfig = new ApiProviderConfig({
    defaultProvider,
    ...config,
  })

  return {
    maxMessageCount: options.maxMessageCount ?? config.maxMessageCount ?? 100,
    budget: options.budget ?? (env.POLKA_BUDGET ? Number.parseFloat(env.POLKA_BUDGET) : undefined) ?? config.budget ?? 10,
    verbose: options.verbose ?? 0,
    config,
    providerConfig,
    agent: options.agent ?? config.agent ?? 'architect',
  }
}
