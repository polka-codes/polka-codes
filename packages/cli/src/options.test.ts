// Generated by polka.codes

import { afterEach, beforeEach, describe, expect, test } from 'bun:test'
import { mkdtempSync, rmSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { Command } from 'commander'

import { ApiProviderConfig } from './ApiProviderConfig'
import { AiProvider } from './getModel'
import { addSharedOptions, parseOptions } from './options'

describe('ApiProviderConfig', () => {
  let testDir: string

  beforeEach(() => {
    testDir = mkdtempSync(join(tmpdir(), 'polkacodes-test-'))
  })

  afterEach(() => {
    rmSync(testDir, { recursive: true, force: true })
  })

  test('getConfigForCommand with command-specific config', () => {
    const config = new ApiProviderConfig({
      defaultProvider: AiProvider.Anthropic,
      providers: {
        [AiProvider.Anthropic]: {
          apiKey: 'test-key',
          defaultModel: 'claude-3-opus',
        },
      },
      commands: {
        chat: {
          provider: AiProvider.DeepSeek,
          model: 'deepseek-chat',
        },
      },
    })

    expect(config.getConfigForCommand('chat')).toMatchSnapshot()
  })

  test('getConfigForCommand falls back to default provider', () => {
    const config = new ApiProviderConfig({
      defaultProvider: AiProvider.Anthropic,
      providers: {
        [AiProvider.Anthropic]: {
          apiKey: 'test-key',
          defaultModel: 'claude-3-opus',
        },
      },
    })

    expect(config.getConfigForCommand('unknown')).toMatchSnapshot()
  })

  test('getConfigForAgent with agent-specific config', () => {
    const config = new ApiProviderConfig({
      defaultProvider: AiProvider.Anthropic,
      providers: {
        [AiProvider.Anthropic]: {
          apiKey: 'test-key',
          defaultModel: 'claude-3-opus',
        },
      },
      agents: {
        coder: {
          model: 'claude-3-sonnet',
        },
      },
    })

    expect(config.getConfigForAgent('coder')).toMatchSnapshot()
  })
})

describe('parseOptions', () => {
  let testDir: string

  beforeEach(() => {
    testDir = mkdtempSync(join(tmpdir(), 'polkacodes-test-'))
  })

  afterEach(() => {
    rmSync(testDir, { recursive: true, force: true })
  })

  test('prioritizes CLI flags over env vars and config', () => {
    const configPath = join(testDir, 'test-config.yml')
    writeFileSync(
      configPath,
      `
defaultProvider: deepseek
defaultModel: deepseek-chat
providers:
  anthropic:
    apiKey: config-key
`,
    )

    const command = new Command()
    addSharedOptions(command)
    const options = command
      .parse([
        'node',
        'test',
        '--api-provider',
        'deepseek',
        '--model',
        'deepseek-chat-32k',
        '--api-key',
        'cli-key',
        '--config',
        configPath,
        '--verbose',
      ])
      .opts()

    const result = parseOptions(options, testDir, testDir, {})
    expect(result.verbose).toBe(1)
    expect(result.providerConfig.getConfigForCommand('chat')).toMatchSnapshot()
  })

  test('handles multiple config files', () => {
    const configPath1 = join(testDir, 'config1.yml')
    const configPath2 = join(testDir, 'config2.yml')

    writeFileSync(
      configPath1,
      `
defaultProvider: anthropic
defaultModel: claude-3-opus
rules:
  - rule1
`,
    )

    writeFileSync(
      configPath2,
      `
defaultModel: claude-3-sonnet
rules:
  - rule2
`,
    )

    const command = new Command()
    addSharedOptions(command)
    const options = command.parse(['node', 'test', '--config', configPath1, '--config', configPath2]).opts()

    const result = parseOptions(options, testDir, testDir, {})
    expect(result.config).toMatchSnapshot()
  })

  test('merges environment variables with config', () => {
    const env = {
      POLKA_API_PROVIDER: AiProvider.DeepSeek,
      POLKA_MODEL: 'deepseek-chat',
      POLKA_API_KEY: 'cli-key',
    }
    const command = new Command()
    addSharedOptions(command)
    const options = command.parse(['node', 'test']).opts()

    const result = parseOptions(options, testDir, testDir, env)
    expect(result.providerConfig.getConfigForCommand('chat')).toMatchSnapshot()
  })

  test('throws error when apiKey provided without provider', () => {
    const command = new Command()
    addSharedOptions(command)
    const options = command.parse(['node', 'test', '--api-key', 'invalid-key']).opts()

    expect(() => parseOptions(options, testDir, testDir, {})).toThrow('Must specify a provider')
  })

  test('handles provider-specific environment variables', () => {
    const env = {
      ANTHROPIC_API_KEY: 'provider-env-key',
      DEEPSEEK_API_KEY: 'deepseek-key',
      OPENROUTER_API_KEY: 'openrouter-key',
    }
    const command = new Command()
    addSharedOptions(command)
    const options = command.parse(['node', 'test']).opts()

    const result = parseOptions(options, testDir, testDir, env)
    expect(result.providerConfig.providers.anthropic?.apiKey).toBe('provider-env-key')
    expect(result.providerConfig.providers.deepseek?.apiKey).toBe('deepseek-key')
    expect(result.providerConfig.providers.openrouter?.apiKey).toBe('openrouter-key')
  })

  test('handles budget from environment', () => {
    const env = {
      POLKA_BUDGET: '500',
    }
    const command = new Command()
    addSharedOptions(command)
    const options = command.parse(['node', 'test']).opts()

    const result = parseOptions(options, testDir, testDir, env)
    expect(result.budget).toBe(500)
  })
})
