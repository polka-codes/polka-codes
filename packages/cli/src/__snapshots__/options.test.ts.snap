// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`ApiProviderConfig getConfigForCommand with command-specific config 1`] = `
{
  "apiKey": undefined,
  "baseUrl": undefined,
  "keyFile": undefined,
  "location": undefined,
  "model": "deepseek-chat",
  "parameters": {},
  "project": undefined,
  "provider": "deepseek",
}
`;

exports[`ApiProviderConfig getConfigForCommand falls back to default provider 1`] = `
{
  "apiKey": "test-key",
  "baseUrl": undefined,
  "keyFile": undefined,
  "location": undefined,
  "model": "claude-3-opus",
  "parameters": {},
  "project": undefined,
  "provider": "anthropic",
}
`;

exports[`ApiProviderConfig getConfigForAgent with agent-specific config 1`] = `
{
  "apiKey": "test-key",
  "baseUrl": undefined,
  "keyFile": undefined,
  "location": undefined,
  "model": "claude-3-sonnet",
  "parameters": {},
  "project": undefined,
  "provider": "anthropic",
}
`;

exports[`parseOptions prioritizes CLI flags over env vars and config 1`] = `
{
  "apiKey": "cli-key",
  "baseUrl": undefined,
  "keyFile": undefined,
  "location": undefined,
  "model": "deepseek-chat-32k",
  "parameters": {},
  "project": undefined,
  "provider": "deepseek",
}
`;

exports[`parseOptions handles multiple config files 1`] = `
{
  "defaultModel": "claude-3-sonnet",
  "defaultProvider": "anthropic",
  "excludeFiles": undefined,
  "providers": {
    "anthropic": {
      "defaultModel": "claude-3-sonnet",
    },
  },
  "rules": [
    "rule1",
    "rule2",
  ],
}
`;

exports[`parseOptions merges environment variables with config 1`] = `
{
  "apiKey": "cli-key",
  "baseUrl": undefined,
  "keyFile": undefined,
  "location": undefined,
  "model": "deepseek-chat",
  "parameters": {},
  "project": undefined,
  "provider": "deepseek",
}
`;
